languages: add koka

mtoohey31 opened the pull request:
This PR adds support for [Koka](https://koka-lang.github.io/koka/doc/index.html). It includes a grammar, highlights, and indents queries.

There isn't currently a complete language server, so I haven't included one (though there is an [unmerged PR](https://github.com/koka-lang/koka/pull/346) which implements one, so if that gets merged I'll create a follow-up PR).

~~The grammar isn't completely up-to-date with the latest version of the language: it doesn't include the new `fip`/`fbip` keywords added in 2.4.2 because the official grammar specification hasn't been updated to include them yet either. I intend to ask about this upstream; depending on when the official grammar spec is updated I may update the tree-sitter grammar's version in this PR before it's merged, or I might do that in a follow-up PR, if that's ok.~~

Edit: nevermind, I found more details on another upstream branch, so this should be fixed now.

Here's a screenshot with the gruvbox theme:

![koka](https://github.com/helix-editor/helix/assets/36740602/ccd5ddda-afc9-4e34-a632-fbf08a48d858)

mtoohey31 commented:
Oh, one other thing I should probably explain: the `tab-width` is 8 but the `unit` is `"  "` because whitespace is significant and the [lexer treats tabs as having a width of 8](https://github.com/koka-lang/koka/blob/1207f26112bee7e58e32c831d647ccfab0fa69da/src/Syntax/Lexer.hs#L43), but the convention is to use two spaces per indent-level.

chtenb commented:
This looks great! I've looked through some example koka code, and noticed the following things, probably something for future improvements.

Is the "scoped" keyword missing?
![image](https://github.com/helix-editor/helix/assets/1030961/9e6c5c6d-e40f-4874-9f94-29c096b15324)

No "local variable" support yet (https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables)
![image](https://github.com/helix-editor/helix/assets/1030961/10b42531-b1d6-4ba2-9622-99394446ffb3)

mtoohey31 commented:
> Is the "scoped" keyword missing?

Yes, thanks for catching that @chtenb! I have a fix on this branch here: https://github.com/mtoohey31/tree-sitter-koka/tree/fix/scoped, would you mind taking a look? The places where `scoped` can be used aren't described in the grammar spec, but it's clearly supported by the compiler. I wasn't quite sure about which of the three `effect` declaration cases to allow it in. The spec line linked in my commit message on that branch seems to indicate it should be allowed for the first two but not for the third, but it's not very clear.

> No "local variable" support yet (https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables)

I've updated the locals queries to handle the example above in d17962b2511c496c81c33cc5b9456ccc9db74eb9. Let me know if you spot anything else that's being missed.

chtenb commented on runtime/queries/koka/highlights.scm:
```diff
@@ -0,0 +1,220 @@
+(puredecl
+  (binder
+    (identifier
+      [(varid) (idop)] @constant)))
```
While technically the `val` declarations could be seen as constants, I think the spirit of this scope is more about compile-time constants such as literal strings/numbers/booleans. They tend to have a distinct color in highlight themes, which may look off when used in such a regular quantity for identifiers. Perhaps the `variable` scope suits this case better. 
On the other hand, it would be nice to have `var` declarations stand out from other (immutable) identifers such as `val`s and function parameters. Perhaps we can find a suitable scope for these? The list of scopes at our disposal is not particularly accomodating unfortunately: https://docs.helix-editor.com/themes.html. What about `variable.other`? Or `special`?

chtenb reviewed (commented):
Are `linear` effects already in the language? If so, that keyword can also be added https://koka-lang.github.io/koka/doc/book.html#sec-linear

I think I looked through all language samples, and there was nothing else that stood out to me.

mtoohey31 commented:
> Are `linear` effects already in the language? If so, that keyword can also be added https://koka-lang.github.io/koka/doc/book.html#sec-linear

I do have `linear` in the grammar and highlights, but I see what you mean in that some of the examples in `test/algeff` put it after the `effect` keyword for some reason? The grammar spec doesn't appear to allow that: https://github.com/koka-lang/koka/blob/875c0cee6ba42e296dfd4f612cb17731c53e236d/doc/spec/grammar/parser.y#L256-L281 though, so I'm not sure which to follow...

> It would be nice to have `var` declarations stand out from other (immutable) identifers such as `val`s and function parameters. Perhaps we can find a suitable scope for these? The list of scopes at our disposal is not particularly accomodating unfortunately: https://docs.helix-editor.com/themes.html. What about `variable.other`? Or `special`?

Yeah it looks like we don't have much selection... `variable.other` doesn't appear to be highlighted in its own right (just `variable.other.member`), and it looks like `special` is the same colour as keywords or functions in a handful of themes. Maybe we should hold off on this for now until there's better options? Kotlin's `val`/`var`, which I believe are similar, don't appear to be highlighted differently.

chtenb commented:
Turns out there is an open issue already https://github.com/helix-editor/helix/issues/6659

Moreover, the rust highlights have an open TODO for this feature: https://github.com/helix-editor/helix/blob/bff7fc8695f1d58b918f94dee7638f8a4a1e4b0e/runtime/queries/rust/highlights.scm#L183

The suggested name is `variable.mut`, but `variable.mutable`, `variable.other.mutable` seem also logical to me. @archseer Can we just add this feature to koka language support, and if so, do you have a strong preference for the naming of this scope?

mtoohey31 commented:
Alright, I've pushed a small fix for ntlappexpr highlighting, and I've added a TODO comment for adding the mutability highlight.

I agree with your suggested solution @chtenb, but I think it should probably be a separate PR cause it'll involve a lot of theme updates, as well as updates to the Rust and Kotlin highlight queries at the very least. Does that sound fair?

If so, I think this PR is ready.

chtenb reviewed (approved):
Sounds good. LGTM

chtenb commented:
I've noticed a minor issue in the indent queries. When entering a new line after something like below, the cursor starts indented.

```
fun parse-input() { many1(parse-line) }
  [] <- cursor starts here
```

chtenb commented:
A small highlighting discrepancy I hadn't noticed before:
![image](https://github.com/helix-editor/helix/assets/1030961/d8cc8892-f8c7-4be5-8b03-6171a7678097)
The dot-notation is a regular function call, so I'd expect the same color.

mtoohey31 commented:
Thanks, I've fixed the function highlighting and resolved the merge conflict. I'm working on the indent query. I can reproduce it but haven't figured out how to fix it yet.

chtenb commented:
I noticed another undocumented keyword missing: `ref`. It's an alias for `reference` it seems. I noticed this because compiler warnings mention this keyword. It looks like Koka has a trend of introducing abbreviations for longer keywords. And the documented grammar is not kept up-to-date with every minor change.

chtenb commented:
> I've noticed a minor issue in the indent queries. When entering a new line after something like below, the cursor starts indented.
> 
> ```
> fun parse-input() { many1(parse-line) }
>   [] <- cursor starts here
> ```

Here is a somewhat similar situation

```koka
fun handle-graph-as-igraph(g : graph, f : () -> <igraph,exn|e> a) : <exn|e> a
  with handler
[]     <---- cursor starts here
```

I think it would make sense to have two or at least one level of indentation here.

mtoohey31 commented:
I've added a couple missing keywords in 9c2af05a1f9ee1d4d35cbd95a34e918b90bcf0a3. You may want to take a look at https://github.com/mtoohey31/tree-sitter-koka/commit/4eef46e8a9aad2b0c0cadb28697a672f76998b62 to see if that's all covered by https://github.com/koka-lang/koka/pull/364.

I'll fix the merge conflicts again in a minute.

Still working on the indentation.

mtoohey31 commented:
Both indentation cases should be fixed now, though the second (the `with handler` one) only has one level of indentation now, not two.

The reason why things break down in that case is that the grammar can't make sense of anything because the incomplete `with handler` followed by EOF is a syntax error, so the queries that would normally match functions, etc. don't work because the function isn't properly parsed. I've made things work somewhat by adding in a specific query for the `(ERROR "fun")` case. The better solution would be to try and get the grammar to localize the syntax error to just the `with handler` statement and still parse the function, but I don't understand tree-sitter well enough to get the grammar to do that.

mtoohey31 commented on runtime/queries/koka/highlights.scm:
```diff
@@ -0,0 +1,237 @@
+  "as"
+  "behind"
+  (externtarget)
+  "forall"
```
`some` is already handled. `exists` does appear in the parser at https://github.com/koka-lang/koka/tree/5703f06c75d9145036940bcb363f40a7db19bb7b, but appears to be unreachable, dead code.

chtenb replied:
Indeed, apologies. I think I missed `some` because it didn't appear beneath forall.

mtoohey31 replied:
No worries, there's so many keywords it's easy to miss.

mtoohey31 commented on runtime/queries/koka/highlights.scm:
```diff
@@ -0,0 +1,237 @@
+[
+  "import"
+  "include"
+  "module"
```
I see this keyword in the docs update, but I'm not sure where it's coming from... it doesn't appear anywhere in any of the `master`, `dev`, or `dev-fbip` branches from the Koka repository.

chtenb replied:
You're right, that was a mistake in the docs update PR

mtoohey31 commented on runtime/queries/koka/highlights.scm:
```diff
@@ -0,0 +1,237 @@
+  "other"
+  "pub"
+  "some"
+] @keyword
```
I don't think there's anything we can do about them. We can't predict how they'll be used, so we can't include them in the grammar. And if they're not in the grammar, you can't include them in the highlights, since they're not a valid query.

If there is a way that you know of though, let me know.

mtoohey31 commented:
Added a few other alternative keywords I found on the Koka `dev` branch. Looks like things are under active development on that branch though, so at some point I'm thinking we have to draw the line and merge things even if it's not 100% up-to-date. We can always create follow-up merge requests to update things as they add features, but at least we'd have things partially working in the meantime.

chtenb commented:
I found a situation where local names are incorrectly colored the same way:
![image](https://github.com/helix-editor/helix/assets/1030961/fc566e5c-ba91-4ea8-b44d-3f967c6b4138)

So <parse|e> is an effect, and this can never be the same thing as the function that is defined, yet `parse` is colored like a function.

mtoohey31 commented:
Thanks for catching all these things @chtenb, those issues should both be fixed now.

chtenb commented:
![image](https://github.com/helix-editor/helix/assets/1030961/be1096d7-32f8-4648-82da-068ffbd9e7c9)

This is an interesting one. Should arguments that are functions also be colored as functions? Notice how `ordering` is colored as a parameter in the parameter list, but as a function in the body.

chtenb commented:
![image](https://github.com/helix-editor/helix/assets/1030961/726cc9c5-e15c-40b8-992b-6217650d5b1d)

I noticed that effect types are not colored like types when they are defined, but they _are_ colored like types when used. I think it makes sense to color always them as types.

chtenb commented:
Here is another comment on indent behavior in `handle` blocks:

```koka
pub effect yield<t>
  ctl yield(token : t) : ()

pub fun is-empty(input : input<t>) : bool
  handle input
    return (_)
      True
    ctl yield(_)
      False
```

If you press enter after `return (_)`, the cursor incorrectly unindents. 
I think that for each of the last 5 lines the given indentation is too short.

chtenb commented:
I found a syntactical construction using `handle` that is valid but seem to break highlighting:
![image](https://github.com/helix-editor/helix/assets/1030961/8285a058-08c3-4a8d-89a0-25198b138668)

For reference, in text form:

```koka
pub effect yield<t>
  ctl yield(token : t) : ()
  
type stream<a,e>
  Stream(next : () -> e maybe<(a, stream<a,e>)>)

fun to-pull-stream(g : () -> <yield<b>|e> ()) : stream<b,e>
  Stream({
    handle g
      return(x) Nothing
      raw ctl yield(b)
        Just((b, Stream(fn() rcontext.resume(()))))
  })
  
pub type parse-result
  ParseOk
  ParseError
```

chtenb commented:
Found a case where keywords are not correctly highlighted, namely in the shorthand effect declaration:

![image](https://github.com/helix-editor/helix/assets/1030961/0a302b93-cf1b-4f74-a8f1-cabf50029b79)

chtenb commented:
One liners with a return type signature:
![image](https://github.com/helix-editor/helix/assets/1030961/797f41e1-d15a-4e0b-a175-e455011a4c12)
Arguably bad practice, but I use it for compact snippets to share on e.g. github..

mtoohey31 commented:
> One liners with a return type signature:
> ![image](https://private-user-images.githubusercontent.com/1030961/292874520-797f41e1-d15a-4e0b-a175-e455011a4c12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2OTgzNDgsIm5iZiI6MTcwMzY5ODA0OCwicGF0aCI6Ii8xMDMwOTYxLzI5Mjg3NDUyMC03OTdmNDFlMS1kMTVhLTRlMGItYTE3NS1lNDU1MDExYTRjMTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjdUMTcyNzI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2FiY2E3YmMxNDVkNmM4ZDNlZWNiMzlhMmEwNTdkM2MyZWE0OWVmNWExYTY0ZGJjZTY3YzY2Nzk1OTg5MWRjOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.OXojJkF5D6iCr1Msa0hHzvldzS0J_8otsRFs120U0RE)
> Arguably bad practice, but I use it for compact snippets to share on e.g. github..

This is invalid according to the grammar spec. Is this used anywhere in the upstream repo? According to the grammar, the only valid ways to do oneliners are by omitting the result type declaration:

```koka
pub fun take-exn(i : () -> exn int) i()
pub fun get() 42
```

...or by explicitly including braces:

```koka
pub fun take-exn(i : () -> exn int) : exn int { i() }
pub fun get() : e int { 42 }
```

mtoohey31 commented:
> ![image](https://private-user-images.githubusercontent.com/1030961/291226618-be1096d7-32f8-4648-82da-068ffbd9e7c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3MDE2MjQsIm5iZiI6MTcwMzcwMTMyNCwicGF0aCI6Ii8xMDMwOTYxLzI5MTIyNjYxOC1iZTEwOTZkNy0zMmY4LTQ2NDgtODJkYS0wNjhmZmJkOWU3YzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjdUMTgyMjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmZlYmI3OGI2ZDMzM2UzYzc5NDU5Mzc0NWQ0MWY3ZjI2NDQ3OWI4NGRjOGQ5ODhlNjM0ZTQyY2E1NzUzNWJkMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.JeIC3iooY6w1FEFlcjCcblA4NyTkxNx5aVEQ3uf90vE)
> 
> This is an interesting one. Should arguments that are functions also be colored as functions? Notice how `ordering` is colored as a parameter in the parameter list, but as a function in the body when it's called, and as a variable again when it's passed in.

Based on how this is handled in other languages, I've chosen to fix this by highlighting it as a `variable.parameter` at call sites too.

chtenb commented:
> This is invalid according to the grammar spec. Is this used anywhere in the upstream repo? 

I haven't seen this used, no. I just noticed that the compiler accepts it, but adding the braces there makes sense for readability purposes. If the grammar spec doesn't allow it either, putting this in the "won't fix" bin is fine I think.

mtoohey31 commented:
> If the grammar spec doesn't allow it either, putting this in the "won't fix" bin is fine I think.

Sounds good!

Sorry about all the :+1: reactions lol, I've been using them to mark off what I've addressed. I believe I've fixed everything as of a26cb8627f7f89e7cb6094becb4eff89262500d5.

chtenb commented:
> Sorry about all the 👍 reactions lol, I've been using them to mark off what I've addressed. I believe I've fixed everything as of [a26cb86](https://github.com/helix-editor/helix/commit/a26cb8627f7f89e7cb6094becb4eff89262500d5).


Cool! No worries, I don't get notifications about reactions or anything

the-mikedavis reviewed (commented):
Just a minor comment about a field that can be removed, otherwise this looks good to me

mtoohey31 commented on languages.toml:
```diff
@@ -3006,3 +3006,16 @@ file-types = ["janet"]
+scope = "source.koka"
+injection-regex = "koka"
+file-types = ["kk"]
+roots = []
```
Great, should be fixed as of 62bda679fe8a67b88d9f9e86984be7e2dbf9011e.
